### ISSUE TITLE
Investigate the differences between a MarkLogic ItemReader and QueryHostBatcher (DMSDK)

### ISSUE THREAD

**Issue by sastafford** [sastafford/marklogic-spring-batch#148]
Sounds like a little functionality overlap between the two.

Is it possible to create an MarkLogic ItemReader that utilizes QueryHostBatcher?
Does it come to a decision of choosing a customer ItemReader OR QueryHostBatcher?

The action is to create an ItemReader that uses QueryHostBatcher and document findings in this issue.

**Comment by venuiyengar** [comment]
Is the intent here is to use QueryHostBatcher instead of DatabaseClient? I believe we can not replace ItemReader all together. I believe I need to focus on MarkLogicItemReader to use QueryHostBatcher. Please clarify.

**Comment by venuiyengar** [comment]
Scott - I investigated this and here is my recommendation:
Reviewed QueryHostBatcher in relation to the DatabaseClient by using release 3 of DM SDK

1. It is possible to use QueryHostBatcher instead of DatabaseClient
2. There may be structural changes need

Let me know if I this is what is intended

**Comment by HemantPuranik** [comment]
I think the original intention for investigating this topic in details is around developing a scalable and efficient solution when reading large volumes of data from MarkLogic. There is apparently some overlap between Spring Batch and QueryHostBatcher
Think about how how to scale Spring Batch ItemReader across multiple parallel threads. Explore Spring Batch partitioning concept and how you would partition large volume of MarkLogic data so that readers can read separate chunks of data in parallel - for example partition data based on host/forest and batch size etc. Then think about how QueryHostBatcher of DMSDK fits in that solution. Does it fit at all? Or would you use the facilities of MarkLogic Java Client API directly? 
DMSDK Query Host batcher is designed to support reading large volumes of data via parallel threads. It internally partitions data and spawn and manages threads that read the data partitions in batches. The customers need to provide listener implementation that captures custom processing over the batches of data provided by QueryHostBatcher. 
To do the same in MarkLogic Spring Batch - you need to provide MarkLogic specific partioner logic to Spring Batch and leverage built-in mechanisms within Spring Batch to scale using multiple parallel threads. Not sure if QueryHostBatcher fits directly fits in this approach since it does a lot more than partitioning of MarkLogic data and kind of overlaps with Spring Batch when it comes to scaling data movement jobs.
One big advantage of Spring Batch though is the ecosystem around it where you have readers and writers to other data stores like RDBMS systems. So in the ideal world we want ecosystem of Spring Batch and MarkLogic specific performance/scalability of QueryHostBatcher in the same framework.
Hope this helps.

**Comment by sastafford** [comment]
@HemantPuranik - thanks for the comments.  In my research of DMSDK so far, I've used the QueryHostBatcher to perform batch transforms and I know that it will definitely come into use when writing to MarkLogic.  I have yet to really dive in to any Spring Batch jobs that read from MarkLogic.  Your talking points above help with some of the test framing.

**Comment by HemantPuranik** [comment]
Here is a demonstration of query partitioning approach using DMSDK QueryHostBatcher. [MarkLogicDMSDKExperiments](https://github.com/HemantPuranik/MarkLogicDMSDKExperiments)

**Comment by venuiyengar** [comment]
I am bit confused about this issue. The way we currently use springbatch with itemreaders, itemwriters and processor is to bring in the data from external sources using itemreaders and process them (transform) and use itemwriters using databaseclient or DMSDK into MarkLogic. 

The issue described here is the reverse process to read data from MarkLogic and why would one use itemReader when the queryHostBatcher provides that functionality. So I am start of stuck with the concept and I am not able to make any progress except to install and understand the demonstration of MarkLogicDMSDKExperiments project.

@sastafford can you clarify this or assign me another issue for my MBO.

**Comment by venuiyengar** [comment]
@sastafford I am going to investigate how to use SpringBatch partitioning with relational database to explore the plumbing involved first.

**Comment by venuiyengar** [comment]
In order to use SpringBatch Partitioning concept we need the following:

1) Implement Partitioner 
This will encapsulate call to queryHostBatcher and the logic for partitioning based on forest, node, etc
The main method to be implemented is public Map<String, ExecutionContext> partition(int gridSize)
The ExecutionContext will have name value pairs based on the partitioning scheme
2) A DTO object to get and set the data that is retrieved from the database
3) Implement PartitionHandler working based on messaging that works at the step level
4) JobBuilderFactory start the the step1 of the job; step1 will step through the slaveStep based on the chunk size and partitioning
There are lot of similarities the way HemantPuranik implemented the concept of partitioning and SpringBatch Partitioning.
The next step for me is to work on an example project to use partitioning for local partitioning with RDBMS using our job approach

**Comment by venuiyengar** [comment]
I have some time to work on how we want to use Spring Partition for itemReader. I would like some input as to what is the requirement for the itemReader.

I was able to understand what the DMSDK project is trying to do with partitions. There are lot of similarities between spring partition and the way this project has been structured. I was not able to understand how we can get min and max settings of host partitions and forest partitions without calling the QueryHostBatcher. In order to write a Spring batch partition these min and max setting are critical (maybe I am missing something here).

For example, if one were to write a ColumnRangePartitioner for a relational database, the partition will be configured based on the gridsize, min and max number of columns from the table or based on the page size. The main method to be overridden for spring batch partition is public Map<String, ExecutionContext> partition(int gridSize) and the execution context holds name value pairs of min and max values for each of the parttions.

@sastafford @HemantPuranik 
If you could elaborate on this that would be helpful so that I can work some example code on this before my next project starts. I have very limited time and I would like to get some input on this. The context of how one would use the spring batch itemReader with partition implementation would be useful here.

**Comment by sastafford** [comment]
Closing this issue.
